backend/src/v2/driver/driver.go

```diff
@@ -1188,43 +1214,235 @@ func resolveInputs(ctx context.Context, dag *metadata.DAG, iterationIndex *int,
+				ctx:           ctx,
+				artifactSpec:  artifactSpec,
+				dag:           dag,
+				pipeline:      pipeline,
+				mlmd:          mlmd,
+				inputs:        inputs,
+				name:          name,
+				artifactError: artifactError,
 			}
-			if taskOutput.GetOutputArtifactKey() == "" {
-				return nil, artifactError(fmt.Errorf("output artifact key is empty"))
+			if err := resolveUpstreamArtifacts(cfg); err != nil {
+				return nil, err
 			}
-			tasks, err := getDAGTasks()
+		default:
+			return nil, artifactError(fmt.Errorf("artifact spec of type %T not implemented yet", t))
+		}
+	}
+	// TODO(Bobgy): validate executor inputs match component inputs definition
+	return inputs, nil
+}
+
+// resolveUpstreamParametersConfig is just a config struct used to store the
+// input parameters of the resolveUpstreamParameters function.
+type resolveUpstreamParametersConfig struct {
+	ctx        context.Context
+	paramSpec  *pipelinespec.TaskInputsSpec_InputParameterSpec
+	dag        *metadata.DAG
+	pipeline   *metadata.Pipeline
+	mlmd       *metadata.Client
+	inputs     *pipelinespec.ExecutorInput_Inputs
+	name       string
+	paramError func(error) error
+}
+
+// resolveUpstreamParameters resolves input parameters that come from upstream
+// tasks. These tasks can be components/containers, which is relatively
+// straightforward, or DAGs, in which case, we need to traverse the graph until
+// we arrive at a component/container (since there can be n nested DAGs).
+func resolveUpstreamParameters(cfg resolveUpstreamParametersConfig) error {
+	taskOutput := cfg.paramSpec.GetTaskOutputParameter()
+	glog.V(4).Info("taskOutput: ", taskOutput)
+	if taskOutput.GetProducerTask() == "" {
+		return cfg.paramError(fmt.Errorf("producer task is empty"))
+	}
+	if taskOutput.GetOutputParameterKey() == "" {
+		return cfg.paramError(fmt.Errorf("output parameter key is empty"))
+	}
+	tasks, err := getDAGTasks(cfg.ctx, cfg.dag, cfg.pipeline, cfg.mlmd, nil)
+	if err != nil {
+		return cfg.paramError(err)
+	}
+
+	// The producer is the task that produces the output that we need to
+	// consume.
+	producer := tasks[taskOutput.GetProducerTask()]
+	glog.V(4).Info("producer: ", producer)
+	currentTask := producer
+	currentSubTaskMaybeDAG := true
+	// Continue looping until we reach a sub-task that is NOT a DAG.
+	for currentSubTaskMaybeDAG {
+		glog.V(4).Info("currentTask: ", currentTask.TaskName())
+		_, outputParametersCustomProperty, err := currentTask.GetParameters()
+		if err != nil {
+			return err
+		}
+		// If the current task is a DAG:
+		if *currentTask.GetExecution().Type == "system.DAGExecution" {
+			// Since currentTask is a DAG, we need to deserialize its
+			// output parameter map so that we can look its
+			// corresponding producer sub-task, reassign currentTask,
+			// and iterate through this loop again.
+			var outputParametersMap map[string]string
+			b, err := outputParametersCustomProperty["Output"].GetStructValue().MarshalJSON()
```

Comment:
Key "Output" only works when a component has a single output. We should loop over the keys here.
```
			for k, value := range outputParametersCustomProperty {
			        var outputParametersMap map[string]string
				b, err := value. GetStructValue().MarshalJSON()
			        if err != nil {
				        return err
			        }
			        json.Unmarshal(b, &outputParametersMap)
				subTaskName := outputParametersMap["producer_subtask"]
				outputArtifactKey := outputParametersMap["output_artifact_key"]
			}
```

Note: This breaks the loop logic.